cmd/lint/runner.go

```diff
@@ -0,0 +1,108 @@
+package lint
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/spf13/cobra"
+
+	"github.com/giantswarm/config-controller/pkg/generator"
+	"github.com/giantswarm/config-controller/pkg/github"
+	"github.com/giantswarm/config-controller/pkg/lint"
+)
+
+const (
+	owner = "giantswarm"
+	repo  = "config"
+)
+
+type runner struct {
+	flag   *flag
+	logger micrologger.Logger
+	stdout io.Writer
+	stderr io.Writer
+}
+
+func (r *runner) Run(cmd *cobra.Command, args []string) error {
+	ctx := context.Background()
+
+	err := r.flag.Validate()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.run(ctx, cmd, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *runner) run(ctx context.Context, cmd *cobra.Command, args []string) error {
+	var store generator.Filesystem
+	{
+		gh, err := github.New(github.Config{
+			Token: r.flag.GitHubToken,
+		})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if r.flag.ConfigVersion != "" {
+			tag, err := gh.GetLatestTag(ctx, owner, repo, r.flag.ConfigVersion)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			store, err = gh.GetFilesByTag(ctx, owner, repo, tag)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+		} else if r.flag.Branch != "" {
+			store, err = gh.GetFilesByBranch(ctx, owner, repo, r.flag.Branch)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	discovery, err := lint.NewDiscovery(store)
+	if err != nil {
+		return microerror.Mask(err)
+	}
```

Comment:
This is a nit, but it looks to me the encapsulation here is a bit off.

For me it could look like

```go
c := lint.Config{
        FilterFuncs: ...
        Store: ...
}

linter, err := lint.New(c)
if err ...

messages := linter.Lint(ctx)
```

Whether we use discovery or iterate over list of functions it's linter implementation detail.